### PR TITLE
[PLATFORM-1505] Created user permissions service

### PR DESCRIPTION
**Pull request by jcellary**
Implemented user permissions service. The service is a client of the core media wiki and as such is a proxy. Also made some fixes to fluent api due to discovered errors and extended it with querying user permissions by user id and not just name

Pinging @sqreek

### CODE REVIEW

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/util/ArrayMapDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/*
+Helper class - MW when returning maps which are empty returns them as [] which causes problems
+during deserialization because jackson tries to set an array to map. One can inherit and use this
+class with proper parameters i.e.
+
+  public static class LongWkDomainDeserializer extends ArrayMapDeserializer {
```
Comment (WikiaTech) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/util/ArrayMapDeserializer.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/domain/User.java [hunk 2]
```diff
@@ -62,4 +63,7 @@ public String getBlockExpiry() {
   public String getGender() {
     return gender;
   }
+
+  @JsonProperty("missing")
+  public String getMissing() { return missing; }
```
Comment (WikiaTech) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/domain/User.java]: MAJOR Move this closing curly brace to the next line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ARightCurlyBraceStartLineCheck)

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java [hunk 3]
```diff
@@ -23,16 +23,20 @@
...
 import java.io.InputStream;
+import java.net.InetAddress;
 import java.net.URISyntaxException;
 import java.util.Objects;
 
 public abstract class MWApiBase implements WikiaChoose, MainModuleOption, MethodOption {
 
+  public static final String USER_IP_HEADER = "X-11717c71abd3ff11b7d1ae0a1716112b";
+
   final Logger logger = LoggerFactory.getLogger(MWApiBase.class);
```
Comment (WikiaTech) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java]: MAJOR Make the "logger" logger private static final and rename it to comply with the format "LOG(?:GER)?". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1312)

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java [hunk 4]
```diff
@@ -126,10 +134,14 @@ protected String buildUrl() throws URISyntaxException {
     return url;
   }
 
+  protected void addHeaders(HttpMessage httpMessage) {
+    if (userIp != null)
```
Comment (WikiaTech) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java]: MAJOR Missing curly brace. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00121)

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/domain/Query.java [hunk 5]
```diff
@@ -68,7 +71,15 @@ public Page getFirstPage() {
     return badRevIds;
   }
 
+  public static class LongWkDomainDeserializer extends ArrayMapDeserializer {
+
+    protected Class getKeyType() { return Long.class; }
```
Comment (WikiaTech) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/domain/Query.java]: MAJOR Move this closing curly brace to the next line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ARightCurlyBraceStartLineCheck)

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/domain/Query.java [hunk 6]
```diff
@@ -68,7 +71,15 @@ public Page getFirstPage() {
...
+
+    protected Class getValueType() { return WkDomain.class; }
```
Comment (WikiaTech) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/domain/Query.java]: MAJOR Move this closing curly brace to the next line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ARightCurlyBraceStartLineCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 6 major

Watch the comments in this conversation to review them.

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java [hunk 7]
```diff
@@ -23,16 +23,21 @@
...
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.InetAddress;
 import java.net.URISyntaxException;
 import java.util.Objects;
 
 public abstract class MWApiBase implements WikiaChoose, MainModuleOption, MethodOption {
 
-  final Logger logger = LoggerFactory.getLogger(MWApiBase.class);
+  public static final String USER_IP_HEADER = "X-11717c71abd3ff11b7d1ae0a1716112b";
```
Comment (sqreek) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java]: What is this exactly for?

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java [hunk 8]
```diff
@@ -122,14 +131,19 @@ protected String buildUrl() throws URISyntaxException {
...
     buildUrl(builder);
     String url = builder.build().toASCIIString();
-    logger.debug(MarkerFactory.getMarker("MWApi url"), url);
+    LOGGER.debug(MarkerFactory.getMarker("MWApi url"), url);
     return url;
   }
 
+  protected void addHeaders(HttpMessage httpMessage) {
+    if (userIp != null) {
+      httpMessage.setHeader(USER_IP_HEADER, userIp.getHostAddress());
```
Comment (sqreek) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java]: is this ok for be always default?

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/domain/Query.java [hunk 9]
```diff
@@ -68,7 +74,14 @@ public Page getFirstPage() {
     return badRevIds;
   }
 
+  public static class LongWkDomainDeserializer extends ArrayMapDeserializer {
```
Comment (sqreek) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/domain/Query.java]: inner class should be at end of class

File: service/user-permissions/build.gradle [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    prod {
+        marathon {
+            useProd = true
+        }
+        resources {
+            instances = 2
+        }
+        environment {
+            CONSUL_TAGS = '["prod", "swagger"]'
+            ENVIRONMENT = "production"
```
Comment (sqreek) [inline service/user-permissions/build.gradle]: prod

File: service/user-permissions/build.gradle [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        }
+    }
+    dev {
+        environment {
+            CONSUL_TAGS = '["testing", "dev", "swagger"]'
+            ENVIRONMENT = "integration"
```
Comment (sqreek) [inline service/user-permissions/build.gradle]: dev

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/UserPermissionsApplication.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+import com.google.inject.Module;
+import com.theoryinpractise.halbuilder.api.RepresentationFactory;
+import com.theoryinpractise.halbuilder.jaxrs.JaxRsHalBuilderSupport;
+import com.theoryinpractise.halbuilder.standard.StandardRepresentationFactory;
+import io.dropwizard.setup.Bootstrap;
+import io.dropwizard.setup.Environment;
+
+public class UserPermissionsApplication extends ConsulGuiceApplication<UserPermissionsConfiguration> {
+
+  public static final RepresentationFactory
```
Comment (sqreek) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/UserPermissionsApplication.java]: can we rid off RepresentationFactory?
We need to update template.

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+    InetAddress inetAddress = parseInetAddress(userIpAddress);
+    UserPermissionsService.UserPermissionAndBlockInfo userPermissionInfo;
+
+    try {
+      userPermissionInfo = userPermissionsService.getUserPermissionsAndBlock(wikiId,
+                                                                             userId,
+                                                                             inetAddress)
+          .orElseThrow(() -> new WebApplicationException(Response.Status.NOT_FOUND));
+    } catch (UserPermissionsServiceException e) {
```
Comment (sqreek) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java]: can we omit this catch, and put Logger inside orElse?

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    } catch (IllegalArgumentException e) {
+      LOGGER.warn(String.format("Could not parse ip address %s", userIp), e);
+      throw new WebApplicationException("Invalid user IP Address provided",
+                                        Response.Status.BAD_REQUEST);
+    }
+
+    return inetAddress;
+  }
+
+  private UserPermissionCheck createUserPermissionCheckResponse(
```
Comment (sqreek) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java]: can we remove this, and use Builder as a replacement?

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+    boolean userHasPermission = userPermissionInfo.getPermissions().contains(requestedPermission);
+
+    return new UserPermissionCheck(userPermissionInfo.getUserId(),
+                                   userPermissionInfo.getUserIp().getHostAddress(),
+                                   requestedPermission,
+                                   userHasPermission,
+                                   userPermissionInfo.isBlocked());
+  }
+
+  private UserPermissions createGetUserPermissionsResponse(
```
Comment (sqreek) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java]: can we remove this and use Builder as a replacement

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+  public Representation getUserPermissions(
+      @ApiParam("User Id") @PathParam("userId") int userId,
+      @ApiParam("Wiki Id") @PathParam("wikiId") int wikiId) {
+
+    UserPermissionsService.UserPermissionInfo userPermissionInfo;
+
+    try {
+      userPermissionInfo = userPermissionsService.getUserPermissions(wikiId, userId)
+          .orElseThrow(() -> new WebApplicationException(Response.Status.NOT_FOUND));
+    } catch (UserPermissionsServiceException e) {
```
Comment (sqreek) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java]: can we remove this catch and move Logger inside orElse()?

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/service/MWUserPermissionsService.java [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  @Override
+  public Optional<UserPermissionInfo> getUserPermissions(int wikiId,
+                                                         int userId
+  ) throws UserPermissionsServiceException {
+
+    Optional<User> user = getUserMWObject(wikiId, userId, Optional.empty());
+
+    return user.map(u -> new UserPermissionInfo(u.getUserId(),
```
Comment (sqreek) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/service/MWUserPermissionsService.java]: can we use builder here?

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/service/MWUserPermissionsService.java [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+  @Override
+  public Optional<UserPermissionAndBlockInfo> getUserPermissionsAndBlock(int wikiId,
+                                               int userId,
+                                               InetAddress userIpAddress
+  ) throws UserPermissionsServiceException {
+
+    Optional<User> user = getUserMWObject(wikiId, userId, Optional.of(userIpAddress));
+
+    return user.map(u -> new UserPermissionAndBlockInfo(u.getUserId(),
```
Comment (sqreek) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/service/MWUserPermissionsService.java]: can we use builder here?

File: service/user-permissions/src/main/resources/user-permissions.yaml [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+      host: ${consul:LOG_HOST:127.0.0.1}
+      port: ${consul:LOG_PORT:1410}
+      fieldNames:
+        message: rawMessage
+        level: rawLevel
+        tags: rawTags
+        context: rawContext
+      customFields:
+        appname: pandora-user-permissions
+        env: ${consul:SERVICE_ENV:dev}
```
Comment (sqreek) [inline service/user-permissions/src/main/resources/user-permissions.yaml]: is this work?

File: service/user-permissions/src/test/java/com/wikia/services/userpermissions/integration/UserPermissionsResourceIntegrationTest.java [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.core.Response;
+
+@Category(IntegrationTest.class)
+public class UserPermissionsResourceIntegrationTest {
```
Comment (sqreek) [inline service/user-permissions/src/test/java/com/wikia/services/userpermissions/integration/UserPermissionsResourceIntegrationTest.java]: can we remove word 'test' from methods, and use Jukito to provide test suits?

File: tools/create-new-service/templates/build.gradle [hunk 21]
```diff
@@ -17,7 +17,7 @@ deployments {
             instances = 2
         }
         environment {
-            CONSUL_TAGS = '["production", "prod", "swagger"]'
+            CONSUL_TAGS = '["prod", "swagger"]'
             ENVIRONMENT = "production"
```
Comment (sqreek) [inline tools/create-new-service/templates/build.gradle]: please change to prod also

File: settings.gradle [hunk 22]
```diff
@@ -61,5 +61,6 @@ include 'service:mobile-config'
 include 'service:mobile-configurations'
 include 'service:static-assets'
 include 'service:user-avatar'
+include 'service:user-permissions'
 include 'service:user-attribute'
-include 'service:user-preference'
\ No newline at end of file
+include 'service:user-preference'
```
Comment (sqreek) [inline settings.gradle]: please try to keep alphabetic order

**Comment by sqreek** [comment]
No more comments for this moment

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java (hunk 13 above)
Comment (jcellary) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java]: Not sure I understand, the exception is thrown by getUserPermissionsAndBlock, how can I avoid the catch? orElse would only help me for null optionals

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java (hunk 7 above)
Comment (jcellary) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java]: It's a header that MW uses for all requests to get information about the ip address of the caller - in our case it is used to check if this ip is blocked, so we set it to the ip that is given as query param to the user permissions service

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java (hunk 8 above)
Comment (jcellary) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java]: not sure I understand - if this information is not provided then header is not set

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/domain/Query.java (hunk 9 above)
Comment (jcellary) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/domain/Query.java]: moved

File: service/user-permissions/build.gradle (hunk 10 above)
Comment (jcellary) [inline service/user-permissions/build.gradle]: fixed

File: service/user-permissions/build.gradle (hunk 11 above)
Comment (jcellary) [inline service/user-permissions/build.gradle]: fixed

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/UserPermissionsApplication.java (hunk 12 above)
Comment (jcellary) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/UserPermissionsApplication.java]: Not sure I understand - I'm using this class in the resource as in the example service. Let's talk offline if there's a new guideline for representation management

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java (hunk 14 above)
Comment (jcellary) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java]: I'm not sure what is the added value of using a builder here - builder is good when you have a lot of parameters and you don't always need to set all of them. In such case you'd have to define a constructor for every possible combination of arguments. In such case builder is a great thing. Here however all arguments are mandatory and the function simply converts the internal class to the response provided by the API. So I'm not convinced if this is a good place to use it.

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java (hunk 15 above)
Comment (jcellary) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java]: Same as above

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java (hunk 16 above)
Comment (jcellary) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java]: I don't think it's possible here - can you provide sample code?

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/service/MWUserPermissionsService.java (hunk 17 above)
Comment (jcellary) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/service/MWUserPermissionsService.java]: Same as above, all arguments are mandatory, so builder IMHO wouldn't help much.

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/service/MWUserPermissionsService.java (hunk 18 above)
Comment (jcellary) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/service/MWUserPermissionsService.java]: as above

File: service/user-permissions/src/main/resources/user-permissions.yaml (hunk 19 above)
Comment (jcellary) [inline service/user-permissions/src/main/resources/user-permissions.yaml]: I'll check I haven't deployed the service to dev yet.

File: tools/create-new-service/templates/build.gradle (hunk 21 above)
Comment (jcellary) [inline tools/create-new-service/templates/build.gradle]: done

File: settings.gradle (hunk 22 above)
Comment (jcellary) [inline settings.gradle]: :+1:

File: service/user-permissions/src/test/java/com/wikia/services/userpermissions/integration/UserPermissionsResourceIntegrationTest.java (hunk 20 above)
Comment (jcellary) [inline service/user-permissions/src/test/java/com/wikia/services/userpermissions/integration/UserPermissionsResourceIntegrationTest.java]: Do we have any service as example?

**Comment by jcellary** [comment]
I fixed some of the issues and commented others. I also made one small extension - added setting of proxy when making requests to MW to icache, so that we don't go through fastly but route the traffic internally. We should make a similar change to the article feeder - I'll talk to the content team.

File: lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java (hunk 7 above)
Comment (jcellary) [inline lib/mediawiki-fluent-api/src/main/java/com/wikia/mwapi/decorator/MWApiBase.java]: needs to be moved to consul

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/UserPermissionsApplication.java (hunk 12 above)
Comment (jcellary) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/UserPermissionsApplication.java]: Need to change as in user preferences

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java (hunk 13 above)
Comment (sqreek) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java]: I understand now
:+1:

File: service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java (hunk 14 above)
Comment (jcellary) [inline service/user-permissions/src/main/java/com/wikia/services/userpermissions/resources/UserPermissionsResource.java]: Ok I'll add a builder with lambook

**Comment by jcellary** [comment]
Resolved all comments and made few more minor tweaks. Merging.

**Pull request merged**